### PR TITLE
Added connection callback, called when the async connection succeeded or failed

### LINKED ISSUE #1464: [FR] Add a callback function for connection resolution readiness
Idea: add a function that would register a callback to be called at the moment when the connection process is resolved.

This registered function will be called at the moment when an asynchronous connection becomes either established or certainly failed. Note that this function shall not be called in case when you use synchronous connection because the process result is then obtained exactly at the moment when the connecting function exits - but only if this is called for a single socket. When you use a synchronous call for a group using `srt_connect_group` - which returns upon the first successful connection and continues the others in background - should still call this callback for every member socket.

This should help in the following situations:

1. When a group member connection fails. There's currently no way at all to determine that an asynchronous connection has failed - it may at best lead to a broken group connection, or it will simply turn out that not all member connections that the application has requested are in the list of the member connections. This callback would be then the only way for an application to determine as to why particular member connection could have failed.

2. When a connection is established, including for a single socket. Even though this thing can be as well determined by the epoll flag or socket state by polling, it might be more comfortable in certain situations. For example, if you want to connect the socket to read from it later, beside setting `SRTO_RCVSYN` you have to subscribe it to epoll with only +[WE] and after you determine [W] readiness on it, you have to resubscribe it with -[W] and +[R]. By using this callback you can simply avoid subscribing this socket in epoll at all and rely on readiness reported by this callback, and only then subscribe it to +[RE], which may simplify the design of the applications using fibre-based socket switching.

How it should work: the callback function should be called:

1. When the async connection is established and the socket becomes ready for transmission. The socket and the value stating that it succeeded should be passed - likely `SRT_SUCCESS`.

2. When the async connection has failed and the socket may only be available to extract information at this last time. In case when this is a single socket, it will be still available outside the callback-called function, but if it is a member of a managed group, this is the last moment when a user may retrieve any error information from it. In this case the error code should be passed as argument, but the user might want to get extra information, such as for example the reject reason, or the extended reject reason from the StreamID field (see #1170 proposal).

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #1464 

This adds the connection callback, which is called in case when the connection that is to be completed in the background is finished, that is, succeeded or failed.

The statement as to whether this could be called in synchronous mode should be left undefined. The only cases when it is intended to be used is:

1. When you use non-blocking mode for a single socket (`SRTO_RCVSYN=false`) - as an alternative for epoll.
2. When you use group connections, in which case single socket connections are done always in the background. In this case it's the only way how you can get error information for every single socket and associate it with the link in your application's link table.

By registerring a callback by `srt_connect_callback`, you install a function that will be called at the moment, when the connection process about the single socket is done. The callback gets:

* the socket in question
* the error code (including `SRT_SUCCESS` if the connection succeeded)
* the target address by `sockaddr*`

The additional reject reason can be obtained by `srt_getrejectreason` on the socket.

### CODE REVIEW

File: srtcore/queue.cpp [hunk 1]
```diff
@@ -970,6 +972,11 @@ void CRendezvousQueue::updateConnStatus(EReadStatus rst, EConnectStatus cst, con
                 i->m_pUDT->m_RejectReason = SRT_REJ_PEER;
             }
             CUDT::s_UDTUnited.m_EPoll.update_events(i->m_iID, i->m_pUDT->m_sPollID, SRT_EPOLL_ERR, true);
+            if (i->m_pUDT->m_cbConnectHook)
+            {
+                CALLBACK_CALL(i->m_pUDT->m_cbConnectHook, i->m_iID,
+                        is_timeout ? SRT_ENOSERVER : SRT_ECONNREJ, i->m_PeerAddr.get());
```
Comment (maxsharabayko) [inline srtcore/queue.cpp:978]: `is_timeout` variable is not necessary.

```suggestion
                CALLBACK_CALL(i->m_pUDT->m_cbConnectHook, i->m_iID,
                        i->m_pUDT->m_RejectReason == SRT_REJ_TIMEOUT ? SRT_ENOSERVER : SRT_ECONNREJ, i->m_PeerAddr.get());
```

File: srtcore/api.cpp [hunk 2]
```diff
@@ -1702,6 +1736,10 @@ int CUDTUnited::connectIn(CUDTSocket* s, const sockaddr_any& target_addr, int32_
        // InvertedGuard unlocks in the constructor, then locks in the
        // destructor, no matter if an exception has fired.
        InvertedLock l_unlocker (s->m_pUDT->m_bSynRecving ? &s->m_ControlLock : 0);
+
+       // record peer address
+       s->m_PeerAddr = target_addr;
+
```
Comment (maxsharabayko) [inline srtcore/api.cpp:1742]: Why is this moved before `startConnect(..)`, and why was it previously after `startConnect(..)`?

Comment (ethouris) [reply srtcore/api.cpp:1742]: Because without this change if `startConnect` fails, this value will never be set.

File: srtcore/queue.cpp [hunk 3]
```diff
@@ -958,8 +958,10 @@ void CRendezvousQueue::updateConnStatus(EReadStatus rst, EConnectStatus cst, con
                 << "). removing from queue");
             // connection timer expired, acknowledge app via epoll
             i->m_pUDT->m_bConnecting = false;
+            bool is_timeout = false;
             if (!is_zero(i->m_tsTTL))
             {
+                is_timeout = true;
```
Comment (maxsharabayko) [inline srtcore/queue.cpp:964]: Or
```suggestion
            const bool is_timeout = !is_zero(i->m_tsTTL);
            if (is_timeout)
            {
```

File: srtcore/api.cpp (hunk 2 above)
Comment (maxsharabayko) [reply srtcore/api.cpp:1742]: We need this value in the connection callback, while previously there was no use of it in case the connection fails, right?

Comment (ethouris) [reply srtcore/api.cpp:1742]: Something like that. Simply, this value wouldn't be set to any different value in case of successful connection, and if it failed, it doesn't matter for the rest of the facility anyway.

**Comment by maxsharabayko** [comment]
- [ ] Update API.md. Note: might be good to merge #1510 first.

**Comment by maxsharabayko** [comment]
- [ ] Fix API.md for listener callback (listeneing - > listening)

> This call installs a callback hook, which will be executed on a socket that is automatically created to handle the incoming connection on the listeneing socket

**Comment by maxsharabayko** [comment]
We have listener callback, that is called on every incoming connection request (conclusion stage) before accepting it.

Now we are going to have connection callback that is called only if the connection has failed, and not called when the connection succeeds.
Maybe it should rather be named "connection failure callback". 🤔

**Comment by maxsharabayko** [comment]
Bonding test to confirm after the fix that the connection callback is called on success as well (extracted from #1416 plus callback added):

[test_bonding.cpp.zip](https://github.com/Haivision/srt/files/5175500/test_bonding.cpp.zip)

**Comment by jeandube** [comment]
Connection callback upon success at the link level is of less interest than failure. connection success is of interest to start sending and this is more for the group socket than the individual links. Connection failure/timeout is of interest at the link level to reissue a link reconnect.

**Comment by ethouris** [comment]
Yes, but it should be there for consistency. You still can distinguish these situations by checking the error code. Also it's not exactly for knowing if to "start sending" - that you know from the group readiness. When the group is connected, it doesn't report any new links connected - you still you the group for sending, no matter how many links it is currently utilizing.

**Comment by maxsharabayko** [comment]
In the case of listener callback, there is a possibility to reject the connection.
The connection callback, as I understand, does not provide this possibility, right?
Maybe it should be rather a "connection failure callback"? Especially given that it is intended only for group connections.

**Note.** [Don't forget](https://github.com/Haivision/srt/pull/1487#issuecomment-685520106) to update the docs with the new function.

**Comment by ethouris** [comment]
It is intended mainly for the failed group member connections, but:
1. If someone already uses it, they can simplify the application architecture by relying on the callback report in every case.
2. The error status check is enough to know if the callback was called in case of a failure.
3. The success callback is also an opportunity to update quickly the application's connection array, or use it as an alternative to epoll-write check in case of a single-socket connections. I've seen already that users generally don't like it and check the socket status periodically instead, which has a disadvantage of having no wait-trigger ability.

File: docs/API-functions.md [hunk 4]
```diff
@@ -652,6 +707,7 @@ typedef struct SRT_GroupMemberConfig_
     int weight;
     SRT_SOCKOPT_CONFIG* config;
     int errorcode;
+	int token;
```
Comment (maxsharabayko) [inline docs/API-functions.md:710]:
```suggestion
    int errorcode;
    int token;
```

File: docs/API-functions.md [hunk 5]
```diff
@@ -688,7 +753,7 @@ typedef struct SRT_SocketGroupData_
     int weight;
     SRT_MEMBERSTATUS memberstate;
     int result;
-
+	int token;
```
Comment (maxsharabayko) [inline docs/API-functions.md:756]: Tab -> spaces

**Review by maxsharabayko — APPROVED** [review]

File: docs/API-functions.md [hunk 6]
```diff
@@ -628,6 +629,60 @@ IMPORTANT: It's not allowed to perform a rendezvous connection to two
...
 
+### srt_connect_callback
+```
+int srt_connect_callback(SRTSOCKET clr, srt_connect_callback_fn* hook_fn, void* hook_opaque);
+```
+
+This call installs a callback hook, which will be executed on a given `clr`
+socket just after the pending connection situation in the background has been
+resolved (that is, when the connection succeeded or failed). Note that this
+function is not guaranteed to be called if the `clr` socket is set blocking
```
Comment (stevomatthews) [inline docs/API-functions.md:640]:
```suggestion
function is not guaranteed to be called if the `clr` socket is set to blocking
```

File: docs/API-functions.md [hunk 7]
```diff
@@ -628,6 +629,60 @@ IMPORTANT: It's not allowed to perform a rendezvous connection to two
...
+This function is mainly intended to be used with group connection. Note that
+even if you use group connection in blocking mode, after the group is considered
+connected, the member connections still continue in background. Also when
+some connections are still pending and some others have failed, the blocking
+call for `srt_connect_group` will not exit until at least one of them succeeds
+or all fail - in such a case those failures also happen only on the background,
+while the connecting function blocks until all connections are resolved. Also,
+when all links fail, you will only get the general error code for the group.
+This mechanism allows you to get individual errors for particular member
+connections.
```
Comment (stevomatthews) [inline docs/API-functions.md:653]:
```suggestion
This function is mainly intended to be used with group connections. Note that
even if you use a group connection in blocking mode, after the group is considered
connected the member connections still continue in background. Also, when
some connections are still pending and others have failed, the blocking
call for `srt_connect_group` will not exit until at least one of them succeeds
or all fail - in such a case those failures also happen only in the background,
while the connecting function blocks until all connections are resolved. 
When all links fail, you will only get a general error code for the group.
This mechanism allows you to get individual errors for particular member
connections.
```

File: docs/API-functions.md [hunk 8]
```diff
@@ -628,6 +629,60 @@ IMPORTANT: It's not allowed to perform a rendezvous connection to two
...
+
+You can also use this mechanism as an alternative method for a single-socket
+connections in non-blocking mode to trigger an action when the connection
+process is finished.
```
Comment (stevomatthews) [inline docs/API-functions.md:657]:
```suggestion
You can also use this mechanism as an alternative method for a single-socket
connection in non-blocking mode to trigger an action when the connection
process is finished.
```

File: docs/API-functions.md [hunk 9]
```diff
@@ -628,6 +629,60 @@ IMPORTANT: It's not allowed to perform a rendezvous connection to two
...
+- Returns:
+
+   * 0, if successful
+   * -1, on error
+
+- Errors:
+
+   * `SRT_EINVPARAM` reported when `hook_fn` is a null pointer
+
+The callback function has the signature as per this type definition:
```
Comment (stevomatthews) [inline docs/API-functions.md:672]:
```suggestion
The callback function signature has the following type definition:
```

File: docs/API-functions.md [hunk 10]
```diff
@@ -675,6 +732,14 @@ parameter's meaning is dependent on the group type:
 The `config` parameter is used to provide options to be set separately
 on a socket for a particular connection  (see [`srt_create_config()`](#srt_create_config)).
 
+The `token` value is inteded to allow the application to easier identify
+particular connection. If you don't use it and leave the default -1 there,
+the library will set unique value for next connections (it's a 32-bit unsigned
+number that will overflow by itself and the -1 value will be skipped).
+The application can as well set a unique value by itself and keep the same
+value for the same connection.
```
Comment (stevomatthews) [inline docs/API-functions.md:740]:
```suggestion
The `token` value is intended to allow the application to more easily identify
a particular connection. If you don't use it and leave the default value of -1,
the library will set a unique value for the next connection (a 32-bit unsigned
number that will overflow by itself; the default value will be skipped).
The application can also set a unique value by itself and keep the same
value for the same connection.
```

File: docs/API-functions.md [hunk 11]
```diff
@@ -917,6 +985,18 @@ it then blocks until at least one connection reports success or if all of them
 fail. Connections that continue in the background after this function exits can
 be then checked status by [`srt_group_data`](#srt_group_data).
 
+As member socket connections are running in the background, for getting known
+that particular connection has succeeded or failed it is recommended
+to use [`srt_connect_callback`](#srt_connect_callback). In this case the
+`token` callback function parameter will be the same as the `token` value used
+for particular item in the `name` connection table.
```
Comment (stevomatthews) [inline docs/API-functions.md:992]:
```suggestion
As member socket connections are running in the background, for determining 
if a particular connection has succeeded or failed it is recommended
to use [`srt_connect_callback`](#srt_connect_callback). In this case the
`token` callback function parameter will be the same as the `token` value used
for the particular item in the `name` connection table.
```

File: docs/API-functions.md [hunk 12]
```diff
@@ -917,6 +985,18 @@ it then blocks until at least one connection reports success or if all of them
...
+
+The `token` value doesn't have any limitations except that -1 value is
+a "trap representation", that is, when set on input it will make the internals
+define a unique value for it. Your application can as well set unique values
+there by itself and in this case the token value will be preserved.
```
Comment (stevomatthews) [inline docs/API-functions.md:997]:
```suggestion
The `token` value doesn't have any limitations except that the -1 value is
a "trap representation", that is, when set on input it will make the internals
define a unique value for the `token`. Your application can also set unique values,
in which case the `token` value will be preserved.
```

**Review by stevomatthews — COMMENTED** [review]
Minor edits.

**Review by stevomatthews — APPROVED** [review]
